sqlglot/generator.py

```diff
@@ -2426,13 +2405,29 @@ def jsonkeyvalue_sql(self, expression: exp.JSONKeyValue) -> str:
         return f"{self.sql(expression, 'this')}{self.JSON_KEY_VALUE_PAIR_SEP} {self.sql(expression, 'expression')}"
 
     def jsonpath_sql(self, expression: exp.JSONPath) -> str:
-        path = generate_json_path(
-            expression.expressions,
-            mapping=self._JSON_PATH_MAPPING,
-            unsupported_callback=self.unsupported,
-            bracketed_key_supported=self.JSON_PATH_BRACKETED_KEY_SUPPORTED,
-        )
-        return f"{self.dialect.QUOTE_START}{path.lstrip('.')}{self.dialect.QUOTE_END}"
+        path = self.expressions(expression, sep="", flat=True).lstrip(".")
+        return f"{self.dialect.JSON_PATH_QUOTE}{path}{self.dialect.JSON_PATH_QUOTE}"
+
+    def json_path_part(self, expression: int | str | exp.JSONPathPart) -> str:
+        if isinstance(expression, exp.JSONPathPart):
+            transform = self.TRANSFORMS.get(expression.__class__)
+            if not callable(transform):
+                self.unsupported(f"Unsupported JSONPathPart type {expression.__class__.__name__}")
+                return ""
+
+            return transform(self, expression)
+
+        if isinstance(expression, int):
+            return str(expression)
+
+        if self.JSON_PATH_SINGLE_QUOTE_ESCAPE:
+            escaped = expression.replace("'", "\\'")
+            escaped = f"'{escaped}'"
+        else:
+            escaped = expression.replace('"', '\\"')
+            escaped = f'"{escaped}"'
```

Comment:
Some context surrounding this section... as mentioned in the previous PR, BQ and Spark escape "unsafe" keys using single quotes, i.e. `"['key with spaces']"`.

I thought of two ways to take this into account in the generator:

- escape the single quotes in line 2425, i.e. `escaped = f"\\'{escaped}\\'"`, in order to leave line 2435 untouched, but that felt wrong because it'd be assuming that `QUOTE_START` and `QUOTE_END` are single quotes.
- change the string quotes being used for the json path string literal, which is what I implemented here; I liked this approach more because it makes the two branches here symmetrical and it feels easier to understand.

Not sure if this different escaping behavior in BQ / Spark can be implemented in a cleaner fashion.. FWIW, I'm kinda ok with how this turned out here, @tobymao any thoughts?

Reply:
why can't we always make quote start the same? that's what we do for regular sql isn't it? so always do single quote to normalize it

Reply:
not really a big fan of this implementation, i'd rather just have one quoting style if possible

Reply:
> why can't we always make quote start the same? that's what we do for regular sql isn't it? so always do single quote to normalize it

We can do that if we escape the single quotes I think, see first bullet. How does that look? So for BQ and Spark you'd have `'$[\'key with spaces \']'`.

Reply:
so sqlglot always generates quotse with single quote ', can we default to double quotes for keys?

Reply:
what's an example of that? just in case this is what you're suggesting, note that this is invalid BQ and will return `NULL` in Spark:

```
'$["key with spaces"]'
```

Moreover, if we did `"$['key with spaces']"` which is basically reverting the quotes, it wouldn't be the same in dialects that use double quotes for identifiers.

Reply:
ok, so if '$["key with spaces"]' is illegal, then just use single quotes and  back slashes as you suggested

Reply:
done https://github.com/tobymao/sqlglot/pull/2920/commits/9c81272791ae57d48262123e051c4e99dd5de01a